### ISSUE TITLE
Dropdowns in a filter, Wrong columns direction, Search field.

### ISSUE THREAD

**Issue by Arteha** [SoftwareBrothers/admin-bro#223]
Hey! :smiley:
### Problems :
- :grinning: The old visual bug in filter component:
![image](https://user-images.githubusercontent.com/22820043/70700545-39177980-1ccb-11ea-8cd7-3afb69abd834.png)
Possible solution is to show the menu above if no space below.
- Columns direction is reversed and the name/id feature is not needed:
![image](https://user-images.githubusercontent.com/22820043/70702402-4c781400-1cce-11ea-9a69-b6c9646627ea.png)

### Suggestions :wink: :
- Could you extend BaseResource with `public async search(value: any, limit: number): Promise<Record[]>` method. Then we can use it for search in Dropdowns.
I already prepared it in `admin-bro-typeorm` Resource class:
![image](https://user-images.githubusercontent.com/22820043/70703951-61a27200-1cd1-11ea-9f77-43296b542964.png)
Usage :smirk: :
![image](https://user-images.githubusercontent.com/22820043/70705114-b1823880-1cd3-11ea-91e4-0b7b203bed40.png)
- Could you make a little delay(about 1.5sec) after keypress in Dropdowns it should be less stressful for database.

**Comment by wojtek-krysiak** [comment]
Hi :) 

1. with the old visual bug I tried to fix that way - we use react-select and I couldn't find a way of changing this direction. But I plan to rewrite the entire UI to [Ant](https://ant.design/) and include its components inside AdminBro - this should fix some issues. But this will be version 2.0
2. This was the design choices we did - we don't force any order except from this:

```
// property-decorator.ts#153
position(): number {
    return this.overrideFromOptions(AvailablePropertyOptions.position, () => {
      if (this.isTitle()) { return -1 }
      if (this.isId()) { return 0 }
      return 100
    })
  }
```

which means unless you provide `position` in PropertyOptions,
* position is set to -1 for title fields, which are `['title', 'name', 'subject', 'email']` or those set by `isTitle` (in PropertyOptions)
* id gets 0
* all other fields gets 100 

...so in order to change the order simply provide "position" in PropertyIOptions.

3. for searching in dropdowns we use regular search - take a look at [search method in api-controller](https://github.com/SoftwareBrothers/admin-bro/blob/master/src/backend/controllers/api-controller.ts#L81).
You can use it as well. Just put the Filter with the given property and that's it.

BTW we will refactor entire Filter interface to make it less magic in version 2.0

**Comment by Arteha** [comment]
1. great.
2. Really really bad idea to forcibly moving columns. Even when if you think they are "necessary"(`['title', 'name', 'subject', 'email']`). Here can be `firstName` of user, `login`, maybe some `code` or whatever. You can't cover all the variants.

When a programmer creates a model, he writes all columns in the dirrection how he wants to see it by default. You can use the first column in model to make it clickable for opening Info. Like it was several versions ago. It was the best solution for these cases:
```typescript
@Entity({ name: "Users" })
export class User extends TypedEntity<Attributes>
{
    @Column({ type: String, length: 32, unique: true, nullable: false })
    public login: string;

    @PrimaryGeneratedColumn()
    public id: number;

    ...
}
```
But all other columns that got poisition 100 are reversed. Probably the reason is inside of sorting algorithm. I temporary patched it in `admin-bro-typeorm` Resouce class:
```typescript
public properties()
{
    // Reverse properties as temporary fix of columns direction.
    // TODO: remove .reverse() when "admin-bro" will be fixed.
    return Object.values(this.propsObject).reverse();
}
```
3. Right now search is not good:
    - You don't use `Like` operator for string/text fields, and it makes people crazy while searching something.
    - However, the best way to give this logic to providers level. Because then it will be possible to use decorators (`@useForSearch()`) in the way like you saw above. I want you to review this issue it again.

I forgot to describe you what does `@UseAsTitle()` decorator.
It gives possibility to generate rich titles for Dropdowns and clicable relations.
Implementation:
```
@Entity({ name: "Volumes" })
export class Volume extends TypedEntity<Attributes>
{
    @PrimaryGeneratedColumn()
    public id: number;

    @Column("integer", { nullable: true })
    public cargoTypeId: number | null;

    @ManyToOne(type => CargoType, cagoType => cagoType.volumes, { onDelete: "CASCADE", eager: true })
    @JoinColumn({ name: "cargoTypeId" })
    public cargoType?: CargoType;

    @Column({ type: "real", nullable: true })
    public from: number | null;

    @Column({ type: "real", nullable: true })
    public to: number | null;

    @OneToMany(type => Price, price => price.volume, { cascade: true })
    public prices?: Array<Price>;

    @UseAsTitle()
    public toString(): string
    {
        return `${this.cargoType || null} [${this.from || "∞"}; ${this.to || "∞"}]`;
    }
}
```
Results:
![image](https://user-images.githubusercontent.com/22820043/71324653-ab653680-24e1-11ea-8b84-50bcc24445d2.png)
![image](https://user-images.githubusercontent.com/22820043/71324761-398dec80-24e3-11ea-9e9b-411ccb5f024d.png)
This decorator can be placed before value/method or even ASYNC method.

**Comment by Arteha** [comment]
I hope you understand me.
Sorry for my perseverance.

Thank you.

**Comment by wojtek-krysiak** [comment]
np. I am just thinking how to move forward with this :) - Let me try to answer. 

2. Regarding this "TitleField": for each resource, we have one "special" property which we treat as a title. This property is "clickable" on the list and it is also presented in autocomplete when searching the relation.

Take a look at this [resource](https://admin-bro-example-app.herokuapp.com/admin/resources/Category/actions/list). _Title_ is treated as a "TitleField" - it is clickable and when you want to add the comment [here](https://admin-bro-example-app.herokuapp.com/admin/resources/Comment/actions/new) - autocomplete works for this field.
So we have to have this "title" field (as an alternative having ID in the autocomplete makes no sense). And along with this choice, we put this "titleField" in front of all other fields (having a clickable field in the middle of a list looked strange). You can also change our "predicted" title field by using "isTitle" PropertyOption (as you mentioned we cannot cover all variants).

Though, you are right by saying that developers usually defines the order of the fields in the code. and this "reverse" you wrote is definitely something we can work on (=I will try to fix that) but we have to put the "TitleField" in front of a list.

3. With the lack of LIKE - I've changed that in admin-bro-typeorm (Like is the default behaviours in both sequelize and mongoose adapters). I am testing this right now on our project and will make PR when done. Until then you can check out the code on our branch: https://github.com/SoftwareBrothers/admin-bro-typeorm. I changed Like in convertFilter but there are other updates as well - related to the latest admin-bro release.

4. Your title implementation looks cool. I will think about how I can migrate this to the admin-bro core.

To summarise I will try to fix this reverse order thing.

**Comment by wojtek-krysiak** [comment]
Hi - I've fixed the order of columns in the latest admin-bro (@1.6.0-beta.5) version and my fork of admin-bro-typeorm (checkout this line: https://github.com/SoftwareBrothers/admin-bro-typeorm/blob/master/src/Property.ts#L49) 

I wanted to make a PR to your remote but found that you made couple of changes which broke automated tests. I will create an issue on your repo
